pkg/mapper/file/mapper.go

```diff
@@ -2,41 +2,60 @@ package file
 
 import (
 	"fmt"
-	"strings"
 
 	"sigs.k8s.io/aws-iam-authenticator/pkg/arn"
 	"sigs.k8s.io/aws-iam-authenticator/pkg/config"
 	"sigs.k8s.io/aws-iam-authenticator/pkg/mapper"
 )
 
 type FileMapper struct {
-	lowercaseRoleMap map[string]config.RoleMapping
-	lowercaseUserMap map[string]config.UserMapping
-	accountMap       map[string]bool
+	roleMap    map[string]config.RoleMapping
+	userMap    map[string]config.UserMapping
+	accountMap map[string]bool
 }
 
 var _ mapper.Mapper = &FileMapper{}
 
 func NewFileMapper(cfg config.Config) (*FileMapper, error) {
 	fileMapper := &FileMapper{
-		lowercaseRoleMap: make(map[string]config.RoleMapping),
-		lowercaseUserMap: make(map[string]config.UserMapping),
-		accountMap:       make(map[string]bool),
+		roleMap:    make(map[string]config.RoleMapping),
+		userMap:    make(map[string]config.UserMapping),
+		accountMap: make(map[string]bool),
 	}
 
 	for _, m := range cfg.RoleMappings {
-		canonicalizedARN, err := arn.Canonicalize(strings.ToLower(m.RoleARN))
+		err := m.Validate()
 		if err != nil {
-			return nil, fmt.Errorf("error canonicalizing ARN: %v", err)
+			return nil, err
 		}
-		fileMapper.lowercaseRoleMap[canonicalizedARN] = m
+		var key string
+		if m.RoleARN != "" {
+			canonicalizedARN, err := arn.Canonicalize(m.RoleARN)
+			if err != nil {
+				return nil, fmt.Errorf("error canonicalizing ARN: %v", err)
+			}
+			key = canonicalizedARN
+		} else {
+			key = m.Key()
+		}
+		fileMapper.roleMap[key] = m
 	}
 	for _, m := range cfg.UserMappings {
-		canonicalizedARN, err := arn.Canonicalize(strings.ToLower(m.UserARN))
+		err := m.Validate()
 		if err != nil {
-			return nil, fmt.Errorf("error canonicalizing ARN: %v", err)
+			return nil, err
+		}
+		var key string
+		if m.UserARN != "" {
+			canonicalizedARN, err := arn.Canonicalize(m.UserARN)
+			if err != nil {
+				return nil, fmt.Errorf("error canonicalizing ARN: %v", err)
+			}
+			key = canonicalizedARN
+		} else {
+			key = m.Key()
```

Comment:
i think i have a gap in my understanding about this part. The UserArn part should not be changing in my impression(since all changes are about adding sso in rolemap). Can you tell more about why this segment of change is neded when UserArn is nil.

Reply:
In a previous commit, I had a top level ARNLikeMapping struct which had UserARNLike and RoleARNLike attributes so I had changes to both. @jaypipes suggested adding the changes in this file to both of those which made both much cleaner.
There's a fair bit of copy/paste code between the two so I though it would be best to keep the implementations similar so they can both benefit from the refactor as part of this PR

Reply:
Thanks for explaining. I think i have a gap in understanding. Can you tell more about how sso relates to roles and users with examples.

Reply:
There's a bit more context in Jay's comment https://github.com/kubernetes-sigs/aws-iam-authenticator/pull/416#pullrequestreview-907912125.

AWS SSO allows federated users to assume roles that it manages. It creates role arns that are long and ugly like: arn:aws:iam::000000000000:role/aws-reserved/sso.amazonaws.com/us-east-1/AWSReservedSSO_MyPermissionSet_1234567890abcde. However the current UX is you have:
```
  mapRoles: |
    - rolearn: arn:aws:iam::000000000000:role/AWSReservedSSO_MyPermissionSet_1234567890abcde
      username: {{SessionName}}
      groups:
        - some-k8s-group
```
Until that suffix changes, which I guess can occur when the SSO permission set changes. If your K8s admins are using SSO roles, this is very bad because another team changing stuff can lock everyone out of k8s.

Also it's confusing because the IAM Role path disappears when it's canonicalized, and there's a few github issues raised due to this.

With the PR as it is currently, the new UX is you create a mapping like this:
```
  mapRoles: |
  - sso:
      permissionSetName: MyPermissionSet
      accountId: "000000000000"
```
and you never need to think about that pesky random suffix or the IAM role path ever again

mapUsers is technically not relevant to this change but I changed it in earlier commits and I think the refactor is valuable. Also I'm sure those having the same method signatures will be useful for generically doing CRD mappings, although I haven't looked into that too much

Reply:
Thank you. I have 2 buckets of question to brainstorm.
-  `mapUsers is technically not relevant to this change but I changed it in earlier commits and I think the refactor is valuable` . I agree, Im aligned on most of the refactorngs. The one im having question about is [this](https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/2f55274303e9d197f6bec4431b6a185943930c35/pkg/mapper/file/mapper.go#L49-L57)
```
		if m.UserARN != "" {
			canonicalizedARN, err := arn.Canonicalize(m.UserARN)
			if err != nil {
				return nil, fmt.Errorf("error canonicalizing ARN: %v", err)
			}
			key = canonicalizedARN
		} else {
			key = m.Key()
		}
```
Under what circumstance is the `m.UserARN` nil?

- Does the sso have to belong in the `mapRoles`? From a user experience perspective could sso matching be a separate stanza altogether? What is your opinion?

Reply:
m.UserARN is a string so it's never nil 😎. But yeah it looks like that block is a relic of the previous implementation so that check is unnecessary and I'll clean it up.

SSO does not have to belong under mapRoles and it didn't initially. @jaypipes requested it be nested under mapRoles and I've requested that you check with him to agree on where you'd like it to sit.
My previous implementation had it as its own stanza so I'd like you folks to agree on it so we can avoid rubber banding the implementation between the two

Reply:
@jyotimahapatra - the comment in question from Jay: https://github.com/kubernetes-sigs/aws-iam-authenticator/pull/416#pullrequestreview-907912125

Reply:
Also, just to share my opinion on this:

I like the idea of having the SSO mapping nested under mapRoles because:
- It's less code to add features to it than add more top-level mappings - particularly in the configmap package and the client CLI commands and tests
- We could make the if/else blocks nicer by having a method for saying what kind of mapping is at play:
```
switch mappingType := rolemapping.Type(); mappingType {
case config.RoleARN:
        // do rolearn specific stuff
case config.SSORole:
        // do sso specific stuff
}
```
- Lastly, it's less rework for me :)

Reply:
@jaypipes which of teh following do you prefer
```
apiVersion: v1
data:
  mapRoles: |
    - rolearn: arn:aws:iam::111122223333:role/my-console-viewer-role
      username: my-console-viewer-role
      groups:
        - eks-console-dashboard-full-access-group
   mapSSO: |
       - permissionSetName: ...
       - accountId: ...
       - userName: ...
       - groups:
         - eks-console-dashboard-restricted-access-group
```
vs
```
apiVersion: v1
data:
  mapRoles: |
    - rolearn: arn:aws:iam::111122223333:role/my-console-viewer-role
      username: my-console-viewer-role
      groups:
        - eks-console-dashboard-full-access-group
    - sso:
      - permissionSetName: ...
      - accountId: ...
      - userName: ...
      - groups:
        - eks-console-dashboard-restricted-access-group
```

Reply:
option 2 - my reasoning above 👍

Reply:
Agree from code perspective we can handle the if elses..what do you think about potential errors when adding a new entry. Say a rolemap exists for unique sso users. In the current approach, does adding an arnlike at the top of the list swallow existing rules? See my example in the previous comment.

Reply:
So technically there's no documentation on the expected behaviour for two matching rolearns. It's undefined behaviour, however the implementation in master will use the last item that matches. E.g.
```
  mapRoles: |
  - rolearn: arn:aws:iam::111122223333:role/my-console-viewer-role
    username: foo
    groups: 
    - foo-group
  - rolearn: arn:aws:iam::111122223333:role/my-console-viewer-role
    username: bar
    groups:
    - bar-group
```
Always maps `arn:aws:iam::111122223333:role/my-console-viewer-role` as username `bar` with groups `bar-group`

The behaviour at the time of writing for this PR is that `sso` blocks and `rolearn` blocks can collide an override each other in the same way. So latest defined item in the array with take precedence. E.g.
```
  - sso:
      permissionSetName: "ViewOnlyAccess"
      accountID: "111122223333"
    username: foo
    groups:
    - foo-group
  - rolearn: arn:aws:iam::111122223333:role/AWSReservedSSO_ViewOnlyAccess_123456789abcde
    username: bar
    groups:
    - bar-group
```
Always maps to username `bar` with groups `bar-group`

My main point here is that this is undefined behaviour, but at least this implementation preserves that

Reply:
@jyotimahapatra having another look at your comment, trying to match the SSO role ARN using the existing rolearn field and using the new SSO matcher is a misconfiguration as the user would be specifying potentially two different k8s users and groups  that they would want a single role mapped against.
Also, there is no "more narrow match" as each rolemapping is inspected and either passes or fails. The first mapping that  succeeds is the one that is applied, and they are tested by doing a range loop over the map that contains the rolemappings, which has an undefined order

Reply:
@jyotimahapatra I prefer nesting the `sso` block within the `mapRoles` blocks, because SSO configuration only relates to roles, never users.